index/generator/library/util.go

```diff
@@ -0,0 +1,356 @@
+
+	// convert revision to referenceName type, ref name could be a branch or tag
+	// if revision is not specified it would be the default branch of the project
+	revision := git.Revision
+	refName := plumbing.ReferenceName(git.Revision)
+
+	if plumbing.IsHash(revision) {
+		// Specifying commit in the reference name is not supported by the go-git library
+		// while doing git.PlainClone()
+		fmt.Printf("Specifying commit in 'revision' is not yet supported.")
+		// overriding revision to empty as we do not support this
+		revision = ""
+	}
+
+	if revision != "" {
+		// lets consider revision to be a branch name first
+		refName = plumbing.NewBranchReferenceName(revision)
+	}
+
+	cloneOptions := &gitpkg.CloneOptions{
+		URL:           git.Url,
+		RemoteName:    git.RemoteName,
+		ReferenceName: refName,
+		SingleBranch:  true,
+		// we don't need history for starter projects
+		Depth: 1,
+	}
+
+	originalPath := ""
+	if git.SubDir != "" {
+		originalPath = path
+		path, err = ioutil.TempDir("", "")
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = gitpkg.PlainClone(path, false, cloneOptions)
+
+	if err != nil {
+
+		// it returns the following error if no matching ref found
+		// if we get this error, we are trying again considering revision as tag, only if revision is specified.
+		if _, ok := err.(gitpkg.NoMatchingRefSpecError); !ok || revision == "" {
+			return err
+		}
+
+		// try again to consider revision as tag name
+		cloneOptions.ReferenceName = plumbing.NewTagReferenceName(revision)
+		// remove if any .git folder downloaded in above try
+		_ = os.RemoveAll(filepath.Join(path, ".git"))
+		_, err = gitpkg.PlainClone(path, false, cloneOptions)
+		if err != nil {
+			return err
+		}
+	}
+
+	// we don't want to download project be a git repo
+	err = os.RemoveAll(filepath.Join(path, ".git"))
+	if err != nil {
+		// we don't need to return (fail) if this happens
+		fmt.Printf("Unable to delete .git from cloned devfile repository")
+	}
+
+	if git.SubDir != "" {
+		err = GitSubDir(path, originalPath,
+			git.SubDir)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// DownloadStackFromGit downloads the stack from a git repo
+func DownloadStackFromGit(git *schema.Git, path string, verbose bool) ([]byte, error) {
+	zipPath := fmt.Sprintf("%s.zip", path)
+
+	// Download from given git url. Downloaded result contains subDir
+	// when specified, if error return empty bytes.
+	if err := DownloadRemoteStack(git, path, verbose); err != nil {
+		return []byte{}, err
+	}
+
+	// Zip directory containing downloaded git repo
+	if err := ZipDir(path, zipPath); err != nil {
+		return []byte{}, err
+	}
+
+	// Read bytes from response and return, error will be nil if successful
+	return ioutil.ReadFile(zipPath)
+}
+
+// DownloadStackFromZipUrl downloads the zip file containing the stack at a given url
+func DownloadStackFromZipUrl(zipUrl string, subDir string, path string) ([]byte, error) {
```

Comment:
The need for extracting is because of `subDir`. Would `subDir` be needed here or just for git repos? If so, I could create an if statement to do `dfutil.HTTPGetRequest` if `subDir` is not specified or `dfutil.GetAndExtractZip` if it is. What is your thoughts on this @johnmcollier?

Reply:
Hmm, yeah let's do that.

Reply:
Was able to use `dfutil.DownloadFile` instead of `dfutil.HTTPGetRequest` (see https://github.com/devfile/library/blob/f0f7e11b17de460c02b59cec4cf031269c25663e/pkg/util/util.go#L1018), keeps consistency with what `dfutil.GetAndExtractZip` uses.